flink-runtime/src/main/java/org/apache/flink/runtime/rpc/resourcemanager/ResourceManager.java

```diff
@@ -116,4 +142,78 @@ public SlotAssignment requestSlot(SlotRequest slotRequest) {
 
 		return new TaskExecutorRegistrationSuccess(new InstanceID(), 5000);
 	}
+
+	/**
+	 * Callback method when current resourceManager is granted leadership
+	 *
+	 * @param newLeaderSessionID unique leadershipID
+	 */
+	void handleGrantLeadership(final UUID newLeaderSessionID) {
+		runAsync(new Runnable() {
+			@Override
+			public void run() {
+				log.info("ResourceManager {} was granted leadership with leader session ID {}", getAddress(), newLeaderSessionID);
+				leaderSessionID = newLeaderSessionID;
+				// confirming the leader session ID might be blocking,
+				leaderElectionService.confirmLeaderSessionID(newLeaderSessionID);
+			}
+		});
+	}
+
+	/**
+	 * Callback method when current resourceManager lose leadership.
+	 */
+	void handleRevokeLeadership() {
+		runAsync(new Runnable() {
+			@Override
+			public void run() {
+				log.info("ResourceManager {} was revoked leadership.", getAddress());
+				jobMasterGateways.clear();
+				leaderSessionID = null;
+			}
+		});
+	}
+
+	/**
+	 * Callback method when an error happened to current resourceManager on leader election
+	 * @param e
+	 */
+	void onLeaderElectionError(final Throwable e) {
+		runAsync(new Runnable() {
+			@Override
+			public void run() {
+				log.error("ResourceManager received an error from the LeaderElectionService.", e);
+				// terminate ResourceManager in case of an error
+				shutDown();
+			}
+		});
+	}
+
+	private class ResourceManagerLeaderContender implements LeaderContender {
+
+		@Override
+		public void grantLeadership(UUID leaderSessionID) {
+			handleGrantLeadership(leaderSessionID);
```

Comment:
I wonder why do you call methods in the main class? Couldn't all code in the methods `handleGrantLeadership`, `revokeLeadership`, `onLeaderElectionError` not simply be implemented in this class? Of course they should still call `runAsync` to execute in the main RPC thread.